crates/deltalake-core/src/operations/write.rs

```diff
@@ -960,4 +1132,95 @@ mod tests {
 
         assert_batches_eq!(&expected, &data);
     }
+
+    #[tokio::test]
+    async fn test_replace_where() {
+        // TODO: add tests for replaceWhere with partition column
+        let tmp_dir = tempdir::TempDir::new("test").unwrap();
+        let tmp_path = std::fs::canonicalize(tmp_dir.path()).unwrap();
+
+        let schema = get_arrow_schema(&None);
+        // let table = DeltaOps::new_in_memory()
+        // let ops = DeltaOps::try_from_uri("/tmp/issue_1957_delta_rs").await.unwrap();
+        // let table = ops
+        let table_path = tmp_path.as_os_str().to_str().unwrap();
+        let ops = DeltaOps::try_from_uri(table_path).await.unwrap();
+
+        let batch = RecordBatch::try_new(
+            Arc::clone(&schema),
+            vec![
+                Arc::new(arrow::array::StringArray::from(vec!["A", "B", "C", "C"])),
+                Arc::new(arrow::array::Int32Array::from(vec![0, 20, 10, 100])),
+                Arc::new(arrow::array::StringArray::from(vec![
+                    "2021-02-02",
+                    "2021-02-03",
+                    "2021-02-02",
+                    "2021-02-04",
+                ])),
+            ],
+        )
+        .unwrap();
+
+        // write some data
+        // let table = DeltaOps(table)
+        let table = ops
+            .write(vec![batch])
+            .with_save_mode(SaveMode::Append)
+            .await
+            .unwrap();
+        assert_eq!(table.version(), 0);
+
+        let batch_add = RecordBatch::try_new(
+            Arc::clone(&schema),
+            vec![
+                Arc::new(arrow::array::StringArray::from(vec!["C"])),
+                Arc::new(arrow::array::Int32Array::from(vec![50])),
+                Arc::new(arrow::array::StringArray::from(vec!["2023-01-01"])),
+            ],
+        )
+        .unwrap();
+
+        let table = DeltaOps(table)
```

Comment:
This test case should be split into two.
1. Happy path
2. Failure due to constraint violation

There should also be a test using a table with partition column since it's a common source of bugs

Reply:
Sure, no problem, was already planning to do both

Reply:
Done both